cpp/src/hash/spark_murmur_hash.cu

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/hashing.hpp>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/table/experimental/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/tabulate.h>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+/**
+ * @brief Computes the hash value of a row in the given table.
+ *
+ * This functor uses Spark conventions for MurmurHash3_32 hashing, which
+ * differs from the standard MurmurHash3_32 implementation. These differences
+ * include:
+ * - specialized tail processing in the hash functor
+ * - specializations for decimal types
+ * - special behavior for serially using the output hash as an input seed for
+ *   the next item
+ * - ignorance of null values: [1], [1, null], and [null, 1] have the same hash
+ *
+ * @tparam hash_function Hash functor to use for hashing elements.
+ * @tparam Nullate A cudf::nullate type describing whether to check for nulls.
+ */
+template <template <typename> class hash_function, typename Nullate>
+class spark_device_row_hasher {
+  friend class cudf::experimental::row::hash::row_hasher<
+    spark_device_row_hasher>;  ///< Allow row_hasher to access private members.
+
+ public:
+  spark_device_row_hasher() = delete;
+
+  /**
+   * @brief Return the hash value of a row in the given table.
+   *
+   * @param row_index The row index to compute the hash value of
+   * @return The hash value of the row
+   */
+  __device__ auto operator()(size_type row_index) const noexcept
+  {
+    return detail::accumulate(
+      _table.begin(),
+      _table.end(),
+      _seed,
+      [row_index, nulls = this->_check_nulls] __device__(auto hash, auto column) {
+        return cudf::type_dispatcher(column.type(),
+                                     element_hasher_adapter<hash_function>{nulls, hash, hash},
+                                     column,
+                                     row_index);
```

Comment:
From my observation, type dispatching inside a device `operator()` will blow the code up significantly. I have little idea about what can we do to avoid it in this situation thus just want to bring more of your attention to this point.

Reply:
It's good to keep an eye out for this. Sometimes it can be prevented. I don't think we can prevent it here because the `spark_device_row_hasher` has to iterate over columns on device and we have to dispatch for each column's type. I believe the use of `cudf::type_dispatcher<cudf::experimental::dispatch_void_if_nested>` below helps to reduce the number of code paths that can access nested types. I copied that from the existing `device_row_hasher` code.

Reply:
I agree, probably unavoidable here unfortunately.